design/secondary-network-ipam.md

```diff
@@ -0,0 +1,175 @@
+
+### Risks and Mitigations
+
+We need to ensure the DHCP address range and static address range do
+not overlap. We should be able to ensure that with careful management
+of the CIDRs.
+
+[inwinstack/ipam](https://github.com/inwinstack/ipam) may not be
+stable or reliable, and we would have to either fix it, fork it, or
+build a replacement.
+
+## Design Details
+
+We need to divide the subnet range for the provisioning network
+between a set of addresses we can use for DHCP and a set for static
+IPs.
+
+We need the installer to allocate IPs for the master nodes as it
+provisions them, and to record that information in the kubernetes
+database so those same IPs are not used for other hosts later.
+
+We need to store the subnet CIDR and existing allocations in the
+kubernetes database somewhere so new IPs can be allocated when hosts
+are provisioned.
+
+The [inwinstack/ipam](https://github.com/inwinstack/ipam) controller
+provides `Pool` and `IP` resources for allocating IPs from address
+ranges. We should evaluate it to see if we can use it for managing the
+IP allocations.
+
+The machine-api-provider-baremetal controller is responsible for
+making decisions about how to configure a host, so it should request
+IPs for secondary networks, assign them to the interfaces, and pass
+the relevant data using the ignition configuration data. It will need
+to create host-specific ignition configuration resources because it
+will be different for each host. It should also set the `Machine` as
+an owner of the `IP` so that the reservation is deleted when the
+`Machine` is deleted.
+
+### Work Items
+
+1. Ensure the IP ranges for secondary networks are captured by the
+   installer and saved to the kubernetes database as `Pool` resources.
+1. Ensure the installer registers the IP allocations for masters.
+1. Ensure the IPAM service is deployed along with the other metal3
+   components.
+1. Update the metal3 machine controller to allocate IPs and create
+   host-specific ignition configurations containing the IPs.
+1. Create image to hold IPAM operator.
+1. Add IPAM operator to metal3 deployment.
+
+### Dependencies
+
+* [inwinstack/ipam](https://github.com/inwinstack/ipam)
+
+### Test Plan
+
+No special requirements
+
+### Upgrade / Downgrade Strategy
+
+Add IPAM operator to deployment configuration
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks [optional]
+
+This further complicates the configuration for the metal3 components
+by adding yet another container/Pod/Deployment.
+
+## Alternatives
+
+We could require an external DHCP and IPAM solution for the secondary
+networks, as we do for the primary network. This complicates
+deployments and requires more services running outside of the cluster
+to know about implementation details of the cluster in order to have
+the external DHCP server pass PXE requests to the dnsmasq instance
+that is part of the metal3 deployment, and which might change hosts
+and IPs if the pod is restarted.
+
```

Comment:
Well, I guess we could do this, but I think it would require writing a new dhcp option backend for ironic and having some kind of agent on the dhcp server to update the configurations.  Currently dnsmasq and ironic need to have access to the same filesystem.

Reply:
Yeah, this is an alternative so the details won't matter as much because we aren't going to do it.

Reply:
Ironic is working with static configuration in this case, and as long as it points back to a vip that ironic can be on, then the world is a happy place. If not, then... we'll need to be able to set configuration.

Reply:
AFAIK ironic only manages the pxe configuration - the dnsmasq configuration is static and in a different container to the ironic conductor?

Reply:
`the dnsmasq configuration is static` - yes, `in a different container to the ironic conductor` - yes (although built from ironic's Dockerfile for some reason). What ironic does modify is PXE configuration, but this is not related to the discussion.

Reply:
Right I see, I was confused.  It's really just the TFTP config we are updating.  So we could separate those if we wanted.